connectors/sources/sharepoint.py

```diff
@@ -0,0 +1,715 @@
+        "parent_web_url": "ParentWebUrl",
+        "_id": "Id",
+        "_timestamp": "LastItemModifiedDate",
+        "creation_time": "Created",
+    },
+    DOCUMENT_LIBRARY: {
+        "title": "Title",
+        "parent_web_url": "ParentWebUrl",
+        "_id": "Id",
+        "_timestamp": "LastItemModifiedDate",
+        "creation_time": "Created",
+    },
+    LIST_ITEM: {
+        "title": "Title",
+        "author_id": "EditorId",
+        "creation_time": "Created",
+        "_timestamp": "Modified",
+    },
+    DRIVE_ITEM: {
+        "title": "Name",
+        "creation_time": "TimeCreated",
+        "_timestamp": "TimeLastModified",
+    },
+}
+
+
+def encode(original_string):
+    """Performs encoding on the name of objects
+    containing special characters in their url, and
+    replaces single quote with two single quote since quote
+    is treated as an escape character
+
+    Args:
+        original_string(string): String containing special characters
+
+    Returns:
+        encoded_string(string): Parsed string without single quotes
+    """
+    encoded_string = urllib.parse.quote(original_string, safe="'")
+    return encoded_string.replace("'", "''")
+
+
+def get_expires_at(expires_in):
+    """Adds seconds to the current utc time.
+
+    Args:
+        expires_in (int): Seconds after which the token will expire.
+    """
+    expires_at = datetime.utcnow() + timedelta(seconds=expires_in)
+    # account for clock skew
+    expires_at -= timedelta(seconds=20)
+    return iso_utc(when=expires_at)
+
+
+def is_expired(expires_at):
+    """Compares the given time with present time
+
+    Args:
+        expires_at (datetime): Time of token expires.
+    """
+    # Recreate in case there's no expires_at present
+    if expires_at is None:
+        return True
+    if not isinstance(expires_at, datetime):
+        expires_at = datetime.fromisoformat(expires_at)
+    if datetime.utcnow() >= expires_at:
+        return True
+    else:
+        return False
+
+
+class SharepointDataSource(BaseDataSource):
+    """Sharepoint"""
+
+    name = "Sharepoint"
+    service_type = "sharepoint"
+
+    def __init__(self, configuration):
+        """Setup the connection to the Sharepoint
+
+        Args:
+            configuration (DataSourceConfiguration): Object of DataSourceConfiguration class.
+        """
+        super().__init__(configuration=configuration)
+        self.is_cloud = self.configuration["is_cloud"]
+        self.site_collections = self.configuration["site_collections"]
+        self.ssl_disabled = self.configuration["ssl_disabled"]
+        self.host_url = self.configuration["host_url"]
+        self.certificate = self.configuration["ssl_ca"]
+        self.enable_content_extraction = self.configuration["enable_content_extraction"]
+        self.retry_count = self.configuration["retry_count"]
+        self.ssl_ctx = False
+        self.session = None
+        self.access_token = None
+        self.token_expires_in = None
```

Comment:
As mentioned in slack, these properties should be handled differently - they are initialised in `ping` and other functions which can cause breakage if `get_docs` is going to be called by different thread/host.